### DISCUSSION TITLE
UI Engineers

### DISCUSSION THREAD

**Discussion by arneschreuder** [COS-301/graduates#8] [General]
This is the discussion board for and with UI Engineers.

**Comment by arneschreuder** [comment]
Please use: 

* https://angular.io
* https://material.angular.io
* https://github.com/angular/flex-layout
* https://www.ngxs.io
* https://angular.io/guide/service-worker-getting-started

**Comment by Ruben-Brits** [comment]
@COS-301/ui-engineers  Good day Guys! I am part of the Architecture "Feature", which means that I am responsible for client side organisation, shared libraries organisation, ie:  making sure the structure of how we approach the UI is well formed. I am creating a WhatsApp group as well for less formal communication.

I also have industry experience with front end which should come in handy. I am looking forward to working with you guys and creating a greatUI experience for the project!

**Comment by MatthewRobinMarsden** [comment]
@COS-301/ui-engineers Hello everyone, I am part of the UI Kit feature group, working as both the designer and the UI engineer, I am currently going to be working on a mock Adobe XD wireframe for the UI Kit that I will then send  to the rest of the designers for a review. Once that has happened, I am going to create the final wireframe (hopefully by mid-next week) and then get to work on converting it to front end.

**Reply by TertiusDeJongh** [reply]
can you maybe send me some of the work-in-progress wireframes? I wish to get an idea of how the overall design is going to look so i can stay consistent when i design something for my team's feature

**Reply by GO-Shoderu** [reply]
Me too

**Comment by MatthewRobinMarsden** [comment]
@COS-301/ui-engineers I would like to ask who is the lead UI engineer. My project manager wishes to discuss topics regarding the UI Kit.

**Reply by RebeccaPillay** [reply]
There are UI Engineers in every (or mostly every) group/feature. If you'd like to speak to someone more involved in the set-up of the general UI libraries etc, perhaps @Ruben-Brits can help. His group is working on the Architecture of the project.

**Reply by MatthewRobinMarsden** [reply]
Thank you for notifying me, I have just done that.

**Reply by rubenbrits** [reply]
Hi, yes my team is doing architecture and I fulfill both the ui engineer and designer roles. More than happy to have a discussion about the UI kit and implementation to ensure we follow a good structure and ensure reusability in the future. We also have to keep in mind that we dont design what cant be catered for in the backend. 
Looking forward to working with you!

**Reply by MatthewRobinMarsden** [reply]
Likewise.

**Comment by Ruben-Brits** [comment]
Hey guys and @arneschreuder ! 
Hope you are doing great.

I've had a few people ask me questions on what UI engineers should currently be busy with. Lol just like you I dont have a ton of info, but what I can tell you is that I am working on finalizing the structure for us. I don't want to give you something now and then ask you to change it in a week and then again.

Good news is I at least have the solution running on my local machine, added a few test features and components, which is a start.

I don't know if there is a point in having a meeting yet? As we can't do much until we have designs. I would say for now if you don't already have a lot of knowledge of Angular and NestJS. Please read up on it and do examples to get familiar with it. Because once we start it is going to move fast I believe.

Please do feel free to message any time if you do need some assistance in any aspect

**Reply by arneschreuder** [reply]
The UI engineers are very dependent on the Designs yes, but they can start at a high level just setting up everything related to the UI. Think about Angular Material Themes, Building of Tailwind, they don't have to implement the theme yet, just some place where the theme can be customised for example. They can start to plan their implementation, what are all the routes going to be, plan a hierarchy of modules and submodules. Learn NGXS!!!, since I assume nobody knows how to use this... this is a must. I can not stress this enough. Our client will be stateful and use NGXS and the stores will make the call,to the API. Components will just subscribe and react to store data.

**Comment by druha4087** [comment]
@arneschreuder the mobile(PWA) feature has been looking into the implementation of PWA scalability for the site, and we were wondering if the alternate UI components that are mobile friendly need to be designed by our team or if we should collaborate with the UI Kit team on this?

**Reply by arneschreuder** [reply]
The mobile/pwa team should not be the ones focussed on the designs. That is why there are designers for each feature as well as a UI Kit team. These designers and UI Kit team and by definition then the UI engineers need to implement the UI in a mobile friendly manner. The PWA team is exactly just that, the PWA team. There is a lot to this, do not underestimate this.

**Comment by GO-Shoderu** [comment]
### **Meeting minutes**

**Reply by MattReed-ZA** [reply]
[UIEngineers_MeetingMinutes_22032022.pdf](https://github.com/COS-301/graduates/files/8327572/UIEngineers_MeetingMinutes_22032022.pdf)

**Comment by arneschreuder** [comment]
@COS-301/ui-engineers Hi UI Engineers. Just checking in if everybody is still ok and alive? There are currently NO work done on the UI yet. This is really really really concerning. What is the deal? Are you guys struggling or what? :) Lets talk, because I am here to help, but I need to know why there are no PRs for the client yet?

**Reply by arneschreuder** [reply]
To put it into perspective. You have basically 2 weeks (18 days) to implement the entire UI. This is really scary guys. On the 18th of April (not confirmed and sure yet) I will create the release branch. If your team's UI component is not in by then, then you will fail and it will DRASTICALLY bring your team's marks down.

**Reply by Ruben-Brits** [reply]
Hi @arneschreuder! With regards to the lack of PR's, I know quite a few of the UI engineers were Designers as well, and most Designs only got completed during last week up until the demo. I have been in contact with the UI engineers and it does seem that everyone is currently busy working on implementing their designs.

**Reply by Ruben-Brits** [reply]
Just to confirm on a architectural level, is it acceptable for the UI features to be in the libs/client directory and we make use of them in the apps/client. Or do you prefer a different structure?

**Comment by ChrisMagerat** [comment]
Good day UI engineers , please note that going forward with this sprint and our PRs, please add any member of the CI/CD team as a assignee and also please remember your meta tags and labels when opening a PR.

**Comment by MatthewRobinMarsden** [comment]
@COS-301/ui-engineers the UI Kit components should be reviewed and uploaded by Wednesday (06/04/2022), reviews for mobile  responsiveness will be done by the PWA team's UI Engineer @SeePeeYou.

**Reply by MatthewRobinMarsden** [reply]
My apologies everyone, because this week is very hectic with work, I have only just received components from my team members to review and thus I do not think we will have components done by the end of the day, we will try and aim for Friday as that is when all the tests and assignments settle down. My apologies once again for the delay.

**Comment by MatthewRobinMarsden** [comment]
### UI Kit Components

Here is a list of UI Kit components and their progress in terms of being pushed to the repo, I will update this as things change, @SeePeeYou this will also be our communication for mobile responsiveness review.

- [x] Header
- [x] Footer
- [x] Buttons
- [x] Drop Down Menu
- [x] Navigation Bar
- [x] Cards
- [x] Tabs

**Reply by SeePeeYou** [reply]
### Mobile PWA assessment of the components

@druha4087, @NasiphiM and I will soon give you feedback about those components as soon as possible using this thread.
Thank you @MatthewRobinMarsden

**Reply by MatthewRobinMarsden** [reply]
@SeePeeYou, @druha4087, @NasiphiM  here is the Cards component for you and your UI Engineers to review.

Link: https://drive.google.com/drive/folders/1Tvu4CICCUVGEv3j-kbJoZOIQAumnZXoe?usp=sharing

**Reply by MatthewRobinMarsden** [reply]
@SeePeeYou, @druha4087, @NasiphiM here is the button and drop down components for you and your UI Engineers to review.

Link: https://drive.google.com/file/d/142PmELJQiQKFtMSsiw1Fhh8e6OvBYbhp/view?usp=sharing

**Reply by MatthewRobinMarsden** [reply]
Here is the footer and header components please review them.

Link: https://drive.google.com/file/d/1J53P4IwxJwHTs_RNPlvJifhEayM2Dpf4/view

**Reply by NasiphiM** [reply]
@MatthewRobinMarsden, after a thorough discussion with my teammates @SeePeeYou and @druha4087, we have agreed that the card components passes enough tests to be mobile friendly and can therefore be published

**Reply by MatthewRobinMarsden** [reply]
Understood will push it then.

**Reply by SeePeeYou** [reply]
@MatthewRobinMarsden after a thorough discussion with my teammates @NasiphiM and @druha4087, we have agreed that the footer components passes enough tests to be mobile friendly and can therefore be published.

**Reply by MatthewRobinMarsden** [reply]
Understood will look to push it then.

**Reply by SeePeeYou** [reply]
@MatthewRobinMarsden The Header for the Mobile PWA works just fine but the header for the Desktop website ids too big, it looks like this

![uiheader](https://user-images.githubusercontent.com/77021097/162619529-6556e400-b4b4-48a2-a372-9f129d6d5947.png)

We will assess it again once it's fixed.

Tests failed:
**CSS Measurements**

- Absolute dimensions should be avoided
- Absolute positions in CSS directives should be avoided as they will not render correctly on all device types

**Reply by MatthewRobinMarsden** [reply]
ok have forwarded it to my group member who did the component.

**Reply by NasiphiM** [reply]
@MatthewRobinMarsden,  @SeePeeYou, @druha4087 and I have looked at the buttons and dropdown menu. We agreed that the dropdown menu is okay to publish although we would like to suggest that you scale down the buttons for mobile devices

**Reply by MatthewRobinMarsden** [reply]
Understood

**Reply by unaizahhhhh** [reply]
@SeePeeYou as discussed earlier about the scaling of the header.
Here is an updated version of the header, please do review and left me know if there needs to be other changes.
https://drive.google.com/file/d/16cW1bKsnXfRIbtYhnSARVc0fW06WjkK_/view?usp=sharing

**Reply by MatthewRobinMarsden** [reply]
@COS-301/ui-engineers Cards have been pushed to the repo.

**Reply by SeePeeYou** [reply]
> @SeePeeYou as discussed earlier about the scaling of the header. Here is an updated version of the header, please do review and left me know if there needs to be other changes. https://drive.google.com/file/d/16cW1bKsnXfRIbtYhnSARVc0fW06WjkK_/view?usp=sharing

As discussed on Discord This version of the Header is still faulty, however, I do approve of the Header you sent on discord. We may still discuss improvements on it but it's sufficient for now. You may publish it. Thank you

**Reply by unaizahhhhh** [reply]
@SeePeeYou here is the a corrected version of  the header and footer, thank you for reviewing
https://drive.google.com/file/d/1gFZGrV_ednXJRI1S8WmTsBqyobgMR-qe/view?usp=sharing

**Reply by SeePeeYou** [reply]
@COS-301/sql Hello, UI Kit team, could you please cater for the Iphone 13 pro max, that will put the maximum width for media at 428

**Reply by MatthewRobinMarsden** [reply]
Understood will do.

**Reply by fffrikkie** [reply]
Do we have an ETA for when the header and footer will be pushed to the repo? And how will we go about accessing/using them once they are pushed?

**Reply by SeePeeYou** [reply]
> Do we have an ETA for when the header and footer will be pushed to the repo? And how will we go about accessing/using them once they are pushed?

Hello, @fffrikkie The UI kit will push to the repo soon, please be advised that it is still under review. the header is buggy for viewports greater than 15"

**Reply by MatthewRobinMarsden** [reply]
The cards component has just been pushed, waiting on @arneschreuder to allow it to be merged if there are no issues.

**Reply by Ruben-Brits** [reply]
@MatthewRobinMarsden Arne does not have to approve and merge every PR apparently from what I heard. If you have a approval and your dev-ops is happy then you can merge

**Reply by arneschreuder** [reply]
Yes please, I am just one human. I can certainly not review and approve everything.

**Reply by MatthewRobinMarsden** [reply]
Understood, the cards component has now been pushed and merged.

**Reply by MatthewRobinMarsden** [reply]
@COS-301/ui-engineers , the header and footer has been pushed to the repo and merged.

**Comment by MatthewRobinMarsden** [comment]
@arneschreuder where exactly do we commit files to the repo for the UI Kit, I mean in terms of the folder structure. E.g: apps or libs?

**Reply by arneschreuder** [reply]
This is documented in the wiki. But the ui kit should be a lib. Like libs/client/shared/. You can have sublibs for different elements like components, services etc. Since this is UI, most of the libs should be have the ui suffix. Like libs/client/shared/components/button/ui

**Reply by MatthewRobinMarsden** [reply]
Ok thank you very much

**Comment by SeePeeYou** [comment]
### Mobile Considerations for ensuring responsiveness throughout the website 

@COS-301/ui-engineers please scan this document and mind the constraints that are applicable to your feature.

Contact @SeePeeYou , @NasiphiM and @druha4087 for questions

[https://docs.google.com/document/d/1me8G6v-TBEvb5iNFCQWCJkkjthGpAn7uQs5Ys6Sf5ao/edit?usp=sharing](https://docs.google.com/document/d/1me8G6v-TBEvb5iNFCQWCJkkjthGpAn7uQs5Ys6Sf5ao/edit?usp=sharing)

**Comment by mvantubbergh** [comment]
@arneschreuder Could you please help me with this problem I am facing? 
When trying to migrate from regular angular project to Nx project, I get these issues when creating an Nx component. The project is an Nx project as visible in the readme file.

The terminal process "C:\WINDOWS\System32\wsl.exe -e npx nx generate @nrwl/angular:component --name=dropdown --project=dropdown --no-interactive" terminated with exit code: 1.

![image](https://user-images.githubusercontent.com/19359163/162969774-97a10d4e-87c7-4272-9f72-525422088042.png)

**Comment by DYL-WF** [comment]
@MatthewRobinMarsden @COS-301/ui-engineers  how do I include the header and footer into my feature. I tried to import the component but it says it was already imported in another ng module? What should be imported exactly?

**Reply by unaizahhhhh** [reply]
Will WhatsApp you.

**Reply by DYL-WF** [reply]
@unaizahhhhh  do you need my number?

**Reply by unaizahhhhh** [reply]
I Found your number on the UI group if I'm not mistaken

**Reply by DYL-WF** [reply]
there are a lot of Dylans

**Reply by MatthewRobinMarsden** [reply]
Everything should be sorted now.
